### PR TITLE
add optional logging capabilities

### PR DESCRIPTION
**Pull request by erikologic**

### CODE REVIEW

File: flagship.go [hunk 1]
```diff
@@ -145,9 +148,10 @@ type featureStore struct {
 	cachedFeatures  models.Features
 	cachedThrottles map[string]*throttleConfigInt
 	store           store
+	logger          *log.Logger
 }
 
-func (s *featureStore) ThrottleAllow(ctx context.Context, key string, hashKey io.Reader) bool {
+func (s *featureStore) throttleAllow(ctx context.Context, key string, hashKey io.Reader) (res bool) {
```
Comment (joerdav) [inline flagship.go]: I'd like to avoid named returns, only for consistencies sake.

File: flagship.go [hunk 2]
```diff
@@ -174,6 +178,15 @@ func (s *featureStore) ThrottleAllow(ctx context.Context, key string, hashKey io
 	return h <= t.Threshold
 
 }
+
+func (s *featureStore) ThrottleAllow(ctx context.Context, key string, hashKey io.Reader) (res bool) {
+	res = s.throttleAllow(ctx, key, hashKey)
+	if s.logger != nil {
+		s.logger.Printf("flagship.ThrottleAllow('%s') --> '%t'", key, res)
```
Comment (joerdav) [inline flagship.go]: I like the idea of this format string, a replacement of `==` for `-->` might make it more familiar, what do you think?

File: flagship.go [hunk 3]
```diff
@@ -174,6 +178,15 @@ func (s *featureStore) ThrottleAllow(ctx context.Context, key string, hashKey io
 	return h <= t.Threshold
 
 }
+
+func (s *featureStore) ThrottleAllow(ctx context.Context, key string, hashKey io.Reader) (res bool) {
```
Comment (joerdav) [inline flagship.go]: I'd like to avoid named returns, only for consistencies sake.

File: flagship.go [hunk 4]
```diff
@@ -184,12 +197,16 @@ func (s *featureStore) GetHash(ctx context.Context, key string, hashKey io.Reade
 	return GetHash(ctx, key, hashKey)
 }
 
-func (s *featureStore) Bool(ctx context.Context, key string) bool {
+func (s *featureStore) Bool(ctx context.Context, key string) (res bool) {
```
Comment (joerdav) [inline flagship.go]: I'd like to avoid named returns, only for consistencies sake.

File: flagship.go [hunk 5]
```diff
@@ -184,12 +197,16 @@ func (s *featureStore) GetHash(ctx context.Context, key string, hashKey io.Reade
...
 	f, _, err := s.fetch(ctx)
 	if err != nil {
 		f = s.cachedFeatures
 	}
-	return f.Bool(key)
+	res = f.Bool(key)
+	if s.logger != nil {
```
Comment (joerdav) [inline flagship.go:206]: I noticed that with the ThrottleAllow method you extracted a separate function, but not for `Bool`. 

Potentially better to keep them similar?

**Review by joerdav — CHANGES_REQUESTED** [review]

File: flagship.go (hunk 5 above)
Comment (erikologic) [reply flagship.go:206]: I didn't wanted to revolutionise the codebase too much
Bool was quite easy to extend and add the logging bit
ThrottleAllow was a different beast

That was my heuristic BTW
I can comply with the request if you confirm this is indeed what we want - please :)

Comment (joerdav) [reply flagship.go:206]: Agree with the thinking, lets keep it separate.

**Comment by erikologic** [comment]
I think this is all sorted now?

BTW sorry for the named returns, it was an artifact of some refactor.

**Review by joerdav — APPROVED** [review]

**Pull request merged**